packages/react-spinbutton/Spec.md

```diff
@@ -2,62 +2,381 @@
+   1. Increment/decrement logic can be handled by comparing `data.value` and `data.prevValue` in the `onChange` callback.
+3. Update `onChange` callback to handle new signature.
+4. Remove `onValidate` callback.
+5. Change ARIA props.
+   1. Change non-standard props to standard ones (e.g., `ariaLabel` to `aria-label`).
+   2. Remove `ariaPositionInSet`, `ariaSetSize`, `ariaValueNow`, `ariaValueText`. The first two are not relevant for a spinbutton and the latter two are internal implementation details managed by the component.
+
 - _Migration from v0_
 
+Not applicable as v0 does not implement this component or one like it.
+
 ## Behaviors
 
-_Explain how the component will behave in use, including:_
+`SpinButton`'s `value` prop is always a number, in contrast to the v8 implementation that gave `value` a string typez. `SpinButton`s manipulate numeric values and making `value` a number aligns it with the other related props: `min`, `max` and `step`. `SpinButton`'s `value` is always displayed as a string which is determined with the `formatter()` function. By default this function stringifies `value` but it can do more sophisticated formatting like apply currency formatting. The inverse of `formatter()` is `parser()` which takes the stringified display value and converts it back to a number. The default implementation calls `parseFloat()` on the display value. `parser()` is called when a user directly types a value into `SpinButton`'s `<input>` element so this value can be converted to a number for computation. If `parser()` returns a non-number value (e.g., `NaN`, `undefined`, `null`) `value` will not be changed.
+
+`formatter()` and `parser()` allow users to hook into `SpinButton`'s display behavior without having to reimplement features like clamping `value` between `min` and `max` which is required when adding [custom suffixes in v8](https://codepen.io/seanms/pen/QWqpQWp).
+
+`SpinButton` does not allow values less than `min` or greater than `max` when these props are specified. When using the step buttons/arrow keys the value will not step outside the min/max bounds. If a value is typed into `<input>` that is outside the min/max range it will be clamped to the relevant bound, that is values less than `min` will be set to `min` and values greater than `max` will be set to `max`.
+
+Aside from min/max range clamping `SpinButton` does not currently implement any input validation.
+
+No error states are currently implemented.
+
+`SpinButton`'s `onChange` callback is invoked every time a change is committed. A change is committed when:
+
+1. A step button is activated
+2. When a user presses the `Up` or `Down` arrow keys while focused on the component.
+3. When a user presses the `Enter` key while focused on the `<input>`
+4. When the `blur` event is fired from the `<input>`
+
+The `onChange` callback is not invoked while a user is focused on the `<input>` and editing the value of the `SpinButton` directly.
```

Comment:
When using a native `<input type="number" /> the change callback is invoked each time the user edits the value of the input field. Why aren't we aligning with the native behaviour here ?

Reply:
This is derived from how it was handled in v8. You can see the spec we followed for v8 in the description of #16137. 

The notion of "intermediate" vs. "committed" values is necessary if there's formatting or validation built in: the user needs to be able to be able to make arbitrary edits to the field before the formatter or validation is run; otherwise they'll just get confused/frustrated when the field keeps on overriding their edits.

If we get rid of value formatting (which IMO we should, if we can persuade design), then we could switch to more similar to native handling and get rid of the notion of intermediate values.

Reply:
I can understand why validation might be an issue, in that case perhaps debouncing might be the way to go there, which is widely used for this purpose.

I can't see why intermediate values are necessary for formatting, is it for a performance reason ? no matter what formatter runs, it should always work for a valid numerical value

Reply:
@ecraig12345  interesting point about intermediate vs committed values

Formatting should work for any numerical value since this is essentially a numeric input.

In terms of validation, the only built in validation I can see in this spec that might be problematic is the min/max clamping. What if we triggered `onChange` event for invalid values ? and simply clamp them back into range on blur ? in that case users can implement controlled input scenarios predictably like other inputs.

It might even be feasible to ignore auto clamping if the component is controlled, since the user should be in full control without the component intervening. In this case the min max props would be handy to apply  aria attributes.

Reply:
The issue with formatting before a value is "committed" is usability. This example will probably help show what I mean. It's a bit of a naive implementation, and it's possible there's a better way to do it that would reduce the issue, but I don't think it will be possible to entirely eliminate without a notion of intermediate/pending value vs. committed value.
https://codesandbox.io/s/spinbutton-formatting-kltvy?file=/src/App.tsx

Reply:
You'd run into a similar issue with validation (which I also do NOT think should be included in the base component aside from clamping to min/max--v8 has more validation and increment/decrement callbacks for "historical reasons," basically in this case that whoever implemented the component thought it was a good idea and we didn't have a dedicated team at the time to ensure contributions were well designed from a code standpoint)

Reply:
And actually, even clamping to min/max while the user is typing could prove problematic--e.g. if they delete the value, what should happen? Or suppose the min is 10 and they select all and type "2" (intending to type "20")?

Reply:
I've forked the sandbox with an alternative: https://codesandbox.io/s/spinbutton-formatting-forked-bk304?file=/src/App.tsx
that can still emit all changes like input fields would normally do. 

For value clamping, this would only happen on blur which is similar to what is being proposed here without getting into the realm of 'committed' and 'non-committed' values.

For formatting, we could just disable the formatting while the input field is being focused, and reenable it on blur. Again it means we are only dealing with numeric values and onChange behaves as normal

Reply:
> For value clamping, this would only happen on blur which is similar to what is being proposed here without getting into the realm of 'committed' and 'non-committed' values.

I'm not sure that calling `onChange` with a value that might be outside the min/max range is a good idea--seems like something that control users likely wouldn't expect and could cause bad behavior in their app (or require extra validation on their end), especially if the value of the SpinButton is being immediately used in the UI somehow. There's also the question of what value to give `onChange` when the user clears the field.

> For formatting, we could just disable the formatting while the input field is being focused, and reenable it on blur. Again it means we are only dealing with numeric values and onChange behaves as normal

So when the user focuses into the field, its text value is updated to only include the number, and then the formatting is added back on blur?

Another important question for all of this is how to avoid creating issues for screen reader users. @smhigley would have a better idea of what to be concerned about there.

Reply:
I also just found out that native `change` behaviour actually behaves like what is specced here, it seems React does the weird thing and makes `onChange` behave exactly the same as `onInput`. 

So I'm no longer against commited/intermediate values, but would prefer a simpler value handling model if it's possible

Reply:
>I'm not sure that calling onChange with a value that might be outside the min/max range is a good idea--seems like something that control users likely wouldn't expect and could cause bad behavior in their app

Not too sure, if they're controlling the value of a numerical input, they should probably do that validation so that we don't impose unnecessary behaviour for what happens when the value goes outside the min/max range

> Another important question for all of this is how to avoid creating issues for screen reader users

Perhaps this might be a motivation to use `div` and `aria-valuenow`... the formatting forces the input to be of type `text` when it's actually handling numeric values... would be willing to hear @smhigley  thoughts on this

Reply:
> Not too sure, if they're controlling the value of a numerical input, they should probably do that validation so that we don't impose unnecessary behaviour for what happens when the value goes outside the min/max range

Given that the SpinButton API exposes `min` and `max` props it seems to me that it should do clamping even when the component is controlled. As a user of this component I would be surprised to get a value outside of the min-max range had I specified those props. Perhaps we need to continue working this API though. Off the top of my head I have a couple thoughts:

1. If `min` and `max` are not specified apply no clamping (or clamp in the range of something like `Number.MIN_VALUE` - `Number.MAX_VALUE` to effectively do no clamping). Currently these props have default values.
2. For more advanced/controlled cases provide a `validator` prop that takes a function that allows component users to apply whatever validation they need. If this prop is provided, ignore `min` and `max`.
    1. Perhaps clamping is always done via this `validator` prop and its default implementation clamps in the range of `min`-`max`? Users can opt out by providing `null` for this prop.

These are not fully formed thoughts and I'd love to discuss this more!

> > Another important question for all of this is how to avoid creating issues for screen reader users
> 
> Perhaps this might be a motivation to use `div` and `aria-valuenow`... the formatting forces the input to be of type `text` when it's actually handling numeric values... would be willing to hear @smhigley thoughts on this

+1

Reply:
> I would be surprised to get a value outside of the min-max range had I specified those props

Yeah I think you're right actually, the validator seems interesting, but I think we want to avoid going down the route of doing form validation.

I tried to make a fiddle with vanilla input and it definitely supports onChange behaviour in this spec (except for Enter, that one is weird): https://jsfiddle.net/wnydu8tc/

Reply:
I'm happy to drop enter though if we pursue the committed/non-committed value approach we will need some event to commit a value. [Commented with some thoughts here](https://github.com/microsoft/fluentui/pull/21081/files#r802190090)